src/parquet/encodings/encoding-test.cc

```diff
@@ -0,0 +1,309 @@
+
+    encode_buffer_ = dst.GetBuffer();
+
+    decoder.SetData(num_values_, encode_buffer_->data(),
+        encode_buffer_->size());
+    int values_decoded = decoder.Decode(decode_buf_, num_values_);
+    ASSERT_EQ(num_values_, values_decoded);
+  }
+
+  void Execute(int nvalues) {
+    InitData(nvalues);
+    GenerateData<T>(nvalues, draws_, &data_buffer_);
+    EncodeDecode();
+    VerifyResults<T>(decode_buf_, draws_, nvalues);
+  }
+
+ private:
+  int num_values_;
+  T* draws_;
+  T* decode_buf_;
+  vector<uint8_t> input_bytes_;
+  vector<uint8_t> output_bytes_;
+  vector<uint8_t> data_buffer_;
+  std::shared_ptr<Buffer> encode_buffer_;
+  std::shared_ptr<ColumnDescriptor> descr_;
+};
+
+TYPED_TEST_CASE(TestPlainEncoding, ParquetTypes);
+
+TYPED_TEST(TestPlainEncoding, BasicRoundTrip) {
+  this->Execute(10000);
+}
+
+// ----------------------------------------------------------------------
+// Dictionary encoding tests
+
+typedef ::testing::Types<Int32Type, Int64Type, Int96Type, FloatType, DoubleType,
+                         ByteArrayType, FLBAType> DictEncodedTypes;
+
+template <typename Type>
+class TestDictionaryEncoding : public ::testing::Test {
+ public:
+  typedef typename Type::c_type T;
+  static constexpr int TYPE = Type::type_num;
+
+  void SetUp() {
+    descr_ = ExampleDescr<T>();
+    if (descr_) {
+      type_length_ = descr_->type_length();
+    }
+  }
+
+  void InitData(int nvalues) {
+    num_values_ = nvalues;
+    input_bytes_.resize(num_values_ * sizeof(T));
+    output_bytes_.resize(num_values_ * sizeof(T));
+    draws_ = reinterpret_cast<T*>(input_bytes_.data());
+    decode_buf_ = reinterpret_cast<T*>(output_bytes_.data());
+  }
+
+  void EncodeDecode() {
+    MemPool pool;
+    DictEncoder<T> encoder(&pool, type_length_);
+
+    dict_buffer_ = std::make_shared<OwnedMutableBuffer>();
+    auto indices = std::make_shared<OwnedMutableBuffer>();
+
+    ASSERT_NO_THROW(
+        {
+          for (int i = 0; i < num_values_; ++i) {
+            encoder.Put(draws_[i]);
+          }
+        });
+    dict_buffer_->Resize(encoder.dict_encoded_size());
+    encoder.WriteDict(dict_buffer_->mutable_data());
+
+    indices->Resize(encoder.EstimatedDataEncodedSize());
+    int actual_bytes = encoder.WriteIndices(indices->mutable_data(),
+        indices->size());
+    indices->Resize(actual_bytes);
+
+    PlainDecoder<TYPE> dict_decoder(descr_.get());
+    dict_decoder.SetData(encoder.num_entries(), dict_buffer_->data(),
+        dict_buffer_->size());
+
+    DictionaryDecoder<TYPE> decoder(descr_.get(), &dict_decoder);
+
+    decoder.SetData(num_values_, indices->data(), indices->size());
+    int values_decoded = decoder.Decode(decode_buf_, num_values_);
+    ASSERT_EQ(num_values_, values_decoded);
+
+    // TODO(wesm): The DictionaryDecoder must stay alive because the decoded
+    // values' data is owned by a buffer inside the DictionaryEncoder. We
+    // should revisit when data lifetime is reviewed more generally.
```

Comment:
I feel we should reuse `InitData` `EnodeDecode` and  `VerifyResults` across various encoders and decoders